packages/core/src/Ponder.ts

```diff
@@ -359,43 +397,79 @@ export class Ponder {
     });
   }
 
+  /**
+   * Very similar to `kill()`, but don't kill the ui service or build service.
+   */
+  private async reload() {
+    this.buildService.clearListeners();
+    this.syncServices.forEach(({ historical, realtime }) => {
+      historical.clearListeners();
+      realtime.clearListeners();
+    });
+    this.serverService.clearListeners();
+    this.indexingService.clearListeners();
+    this.common.metrics.resetMetrics();
+    this.syncGatewayService.clearListeners();
+
+    await Promise.all([
+      await this.indexingService.kill(),
+      await this.serverService.kill(),
+      await this.common.telemetry.kill(),
+    ]);
+
+    await Promise.all(
+      this.syncServices.map(async ({ realtime, historical }) => {
+        await realtime.kill();
+        await historical.kill();
+      }),
+    );
+
+    await this.indexingStore.kill();
+    await this.syncStore.kill();
+  }
+
   private registerServiceDependencies() {
-    this.buildService.on("newConfig", async () => {
-      this.common.logger.info({
-        service: "build",
-        msg: "Detected change in ponder.config.ts",
-      });
-      this.common.logger.info({
-        service: "app",
-        msg: "Reloading ponder, killing and restarting services",
-      });
-      await this.kill();
+    this.buildService.on("newConfig", async ({ config }) => {
+      if (config) {
+        this.common.errors.hasUserError = false;
+        this.common.logger.info({
+          service: "app",
+          msg: "Reloading ponder with new config",
+        });
 
-      // Clear all listeners. Will be added back in setup.
-      this.buildService.clearListeners();
-      this.syncServices.forEach(({ historical, realtime }) => {
-        historical.clearListeners();
-        realtime.clearListeners();
-      });
-      this.syncGatewayService.clearListeners();
-      this.serverService.clearListeners();
-      this.indexingService.clearListeners();
-      this.common.metrics.resetMetrics();
-
-      await this.setup();
-      // NOTE: We know we are in dev mode if the build service is receiving events. Build events are disabled in production.
-      await this.dev();
+        // Clear all listeners. Will be added back in setup.
+        await this.reload();
+
+        await this.setupPrepare({ config });
+        // NOTE: We know we are in dev mode if the build service is receiving events. Build events are disabled in production.
+        await this.dev();
+      } else {
```

Comment:
Might be nice to add a comment here:
```
If the build service emits the "newConfig" event with undefined, it means there was an error
while building or validating the config on a hot reload.
```

Also - do we need to reset metrics here like we're doing in the `admin:reload` handler? If so, might make sense to handle metric resets within the `kill` function of the sync and indexing services.